pkg/agent/externalnode/external_node_controller_test.go

```diff
@@ -0,0 +1,573 @@
+	c.policyBypassRules = []agentconfig.PolicyBypassRule{
+		{
+			Direction: "ingress",
+			Protocol:  "tcp",
+			CIDR:      "10.20.0.0/16",
+			Port:      233,
+		},
+		{
+			Direction: "engress",
+			Protocol:  "udp",
+			CIDR:      "10.30.0.0/16",
+			Port:      244,
+		},
+	}
+	_, cidrIngress, _ := net.ParseCIDR("10.20.0.0/16")
+	_, cidrEgress, _ := net.ParseCIDR("10.30.0.0/16")
+	mockOFClient.EXPECT().InstallPolicyBypassFlows(binding.ProtocolTCP, cidrIngress, uint16(233), true).Times(1)
+	mockOFClient.EXPECT().InstallPolicyBypassFlows(binding.ProtocolUDP, cidrEgress, uint16(244), false).Times(1)
+	c.reconcile()
+}
+
+func TestAddExternalNode(t *testing.T) {
+	defer mockGetIPNetDeviceFromIP([]mockGetIPNetDeviceFromIPParam{
+		{&net.Interface{Name: "intf1"}, nil},
+	})()
+	c := newExternalNodeController(t)
+	externalNode := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{Name: "eeName", Namespace: "ns1", Labels: map[string]string{"en": "vm1"}},
+		Spec: v1alpha1.ExternalNodeSpec{
+			Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.2"}}},
+		},
+	}
+	mockIfaceStore.EXPECT().GetInterfaceByName("intf1").Return(&intf1, true)
+	mockOVSBridgeClient.EXPECT().GetPortData(portUUID1, "intf1").Return(portData1, nil)
+	expectedInfo := map[string]interface{}{
+		"uplink-name":      "uplinkName",
+		"entity-name":      "eeName",
+		"antrea-type":      "host",
+		"entity-namespace": "externalEntityNamespace",
+		"ip-address":       "1.1.1.2",
+		"uplink-port":      portUUID2,
+	}
+	mockOVSBridgeClient.EXPECT().SetPortExternalIDs(intf1.InterfaceName, expectedInfo)
+	iface := &interfacestore.InterfaceConfig{
+		InterfaceName: intf1.InterfaceName,
+		Type:          interfacestore.ExternalEntityInterface,
+		OVSPortConfig: &interfacestore.OVSPortConfig{
+			PortUUID: intf1.PortUUID,
+			OFPort:   intf1.OFPort,
+		},
+		EntityInterfaceConfig: &interfacestore.EntityInterfaceConfig{
+			EntityName:      "eeName",
+			EntityNamespace: intf1.EntityNamespace,
+			UplinkPort: &interfacestore.OVSPortConfig{
+				PortUUID: intf1.UplinkPort.PortUUID,
+				OFPort:   intf1.UplinkPort.OFPort,
+			},
+		},
+		IPs: []net.IP{net.ParseIP("1.1.1.2")},
+	}
+	mockIfaceStore.EXPECT().AddInterface(iface)
+	c.addExternalNode(externalNode)
+	assert.Equal(t, externalNode, c.syncedExternalNode)
+}
+
+func TestEnqueueExternalNodeUpdate(t *testing.T) {
+	c := newExternalNodeController(t)
+	externalNode1 := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{Name: "vm1", Namespace: "ns1", Labels: map[string]string{"en": "vm1"}},
+		Spec: v1alpha1.ExternalNodeSpec{
+			Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.2"}}},
+		},
+	}
+	externalNode2 := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{Name: "vm1", Namespace: "ns1", Labels: map[string]string{"en": "vm1"}},
+		Spec: v1alpha1.ExternalNodeSpec{
+			Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.2"}}},
+		},
+	}
```

Comment:
Just type externalNode1 twice when it tests no update case is more clear and less duplicate.

Reply:
updated